assets/src/components/conversations/ConversationsProvider.tsx

```diff
@@ -207,106 +126,29 @@ export class ConversationsProvider extends React.Component<Props, State> {
-    if (this.socket && this.socket.disconnect) {
-      logger.debug('Existing socket:', this.socket);
-      this.socket.disconnect();
-    }
-
-    this.socket = new Socket(SOCKET_URL, {
-      params: {token: API.getAccessToken()},
-    });
-
-    this.socket.connect();
-    // TODO: attempt refreshing access token?
-    this.socket.onError(
-      throttle(
-        () =>
-          logger.error('Error connecting to socket. Try refreshing the page.'),
-        30 * 1000 // throttle every 30 secs
-      )
-    );
-
-    if (this.channel && this.channel.leave) {
-      logger.debug('Existing channel:', this.channel);
-      this.channel.leave(); // TODO: what's the best practice here?
-    }
-
-    // TODO: If no conversations exist, should we create a conversation with us
-    // so new users can play around with the chat right away and give us feedback?
-    this.channel = this.socket.channel(`notification:${accountId}`, {});
-
-    // TODO: rename to message:created?
-    this.channel.on('shout', (message) => {
-      // Handle new message
-      this.handleNewMessage(message);
-    });
-
-    // TODO: fix race condition between this event and `shout` above
-    this.channel.on('conversation:created', ({id, conversation}) => {
-      // Handle conversation created
-      this.handleNewConversation(id);
-    });
-
-    // TODO: can probably use this for more things
-    this.channel.on('conversation:updated', ({id, updates}) => {
-      // Handle conversation updated
-      this.debouncedConversationUpdate(id, updates);
-    });
-
-    this.channel.on('presence_state', (state) => {
-      this.handlePresenceState(state);
-    });
-
-    this.channel.on('presence_diff', (diff) => {
-      this.handlePresenceDiff(diff);
-    });
-
-    this.channel
-      .join()
-      .receive('ok', (res) => {
-        logger.debug('Joined channel successfully', res);
-      })
-      .receive('error', (err) => {
-        logger.error('Unable to join', err);
-        // TODO: double check that this works (retries after 10s)
-        setTimeout(() => this.joinNotificationChannel(accountId), 10000);
-      });
-  };
-
-  handlePresenceState = (state: PhoenixPresence) => {
+  handlePresenceInit = (state: PhoenixPresence) => {
     this.setState({presence: state});
   };
 
   handlePresenceDiff = (diff: PresenceDiff) => {
-    const {joins, leaves} = diff;
-    const {presence} = this.state;
-
-    const withJoins = updatePresenceWithJoiners(joins, presence);
-    const withLeaves = updatePresenceWithExiters(leaves, presence);
-    const combined = {...withJoins, ...withLeaves};
-    const latest = Object.keys(combined).reduce((acc, key: string) => {
-      if (!combined[key]) {
-        return acc;
-      }
-
-      return {...acc, [key]: combined[key]};
-    }, {} as PhoenixPresence);
-
-    this.setState({presence: latest});
+    this.setState({
+      presence: updatePresenceWithDiff(this.state.presence, diff),
```

Comment:
By refactoring presence related logic into its own file, I was able to extract away everything and only expose this one method for diffing presence.